samples/sample_definition_and_criteria.rst

```diff
@@ -0,0 +1,104 @@
+===============
+
+1. Samples must not be used as a testcase.
+++++++++++++++++++++++++++++++++++++++++++
+  * The primary purpose of a sample is to provide a reference to the user.
+  * Samples must not use Zephyr's testing framework.
+
+    * Must not use :kconfig:option:`CONFIG_ZTEST` or :kconfig:option:`CONFIG_ZTEST_NEW_API`.
+    * Must not use zasserts in samples.
+
+  * If a sample can provide output that can be verified, then output should be evaluated against
+    expected value to have some level of testing for the sample itself.
+  * Samples are optional.
+
+2. Samples should target multiple platforms and architectures.
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+  * Minimize the use of ``platform_allow`` and architecture filters as it limits the scope
+    of testing to the mentioned platforms and architectures. Reference: `Testcase Docs`_
+  * Make use of ``integration_platforms`` to limit the scope when there are timing or
+    resource constraints.
+  * Make the sample as generic as possible. Avoid making a sample platform specific unless it is
+    for particular hardware.
+
+  .. _Testcase Docs: https://docs.zephyrproject.org/latest/develop/test/twister.html?highlight=test%20runner#test-cases.
+
+3. Twister should be able to build every sample.
+++++++++++++++++++++++++++++++++++++++++++++++++
+  * Every sample must have a YAML file.
+
+    **For example:**
+
+    .. code-block:: yaml
+
+      tests:
+        sample.kernel.cond_var:
+          integration_platforms:
+            - native_posix
+          tags: kernel condition_variables
+          harness: console
+          harness_config:
+            type: one_line
+            regex:
+              - ".*Waited and joined with 3 threads"
+
+  **Guidelines for Samples:**
+    * Minimize the use of ``platform_allow`` and architecture filters.
+    * Do not mark the test as build only unless necessary.
+    * Any test case-specific configuration options are added using ``extra_args`` or
+      ``extra_configs`` options in the YAML file. The prj.conf file should have the in-common
+      configuration options.
+    * The sample itself can be evaluated using multiple configurations in the sample's YAML file.
+      For example, the sample is used to run with different logging modes using multiple scenarios
+      in samples/subsys/logging/syst.
+    * Sample output can be validated leveraging the ``harness_config`` regex option,
+      wherever applicable.
+
+4. A sample should provide sufficient coverage of a subsystem, feature, or module.
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+  **DO's:**
+    * Cover the most common and important use cases of the functionality.
+    * Keep the code simple and easy to read. Example: samples/philosophers.
+
+  **DONT's:**
+    * Samples must not test the negative or edge case behaviors.
+    * Must not be unit tests.
+    * Must not incorporate more than one subsystem or feature.
```

Comment:
I don't think it's realistic to limit samples to only incorporate one subsystem or feature.

Reply:
I find this quite difficult as well, even accepting a distinction between samples and ‘demos’.

Reply:
agree, samples should also show how things work together and cant be strictly limited to 1 single subsystem/feature and be useful at the same time. While the main focus needs to go to one subsystem, hybrid samples showcasing multiple subsystem are more than fine.

Reply:
@henrikbrixandersen Please refer to this discussion where a new "demos" category is proposed in the related RFC: [https://github.com/zephyrproject-rtos/zephyr/pull/47978#discussion_r956475329](https://github.com/zephyrproject-rtos/zephyr/pull/47978#discussion_r956475329).

Reply:
@henrikbrixandersen Updated the definition of samples and so removed this point from here. Thanks!